core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -190,95 +145,89 @@ public Object visitLockControlCommand(TxInvocationContext ctx, LockControlComman
       if (!ctx.isInTxScope())
          throw new IllegalStateException("Locks should only be acquired within the scope of a transaction!");
 
-      boolean skipLocking = hasSkipLocking(command);
-      if (skipLocking) {
+      if (command.isUnlock()) { //TODO double check unlock. it isn't used anymore!
+         if (ctx.isOriginLocal()) {
+            throw new AssertionError(
+                  "There's no advancedCache.unlock so this must have originated remotely.");
+         }
          return false;
       }
 
-      // First go through the distribution interceptor to acquire the remote lock - required by DLD.
-      // Only acquire remote lock if multiple keys or the single key primary owner is not the local node.
-      if (ctx.isOriginLocal()) {
-         final boolean isSingleKeyAndLocal =
-               !command.multipleKeys() && cdl.getCacheTopology().getDistribution(command.getSingleKey()).isPrimary();
-         boolean needBackupLocks = !isSingleKeyAndLocal || isStateTransferInProgress();
-         if (needBackupLocks && !command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-            LocalTransaction localTx = (LocalTransaction) ctx.getCacheTransaction();
-            if (localTx.getAffectedKeys().containsAll(command.getKeys())) {
-               if (trace)
-                  log.tracef("Already own locks on keys: %s, skipping remote call", command.getKeys());
-               return true;
-            }
-         } else {
-            if (trace)
-               log.tracef("Single key %s and local, skipping remote call", command.getSingleKey());
-            return localLockCommandWork(ctx, command);
+      if (hasSkipLocking(command) || areAllKeysAlreadyLocked(ctx, command.getKeys())) {
+         //lock control with skip locking?? that is weird...
+         if (trace) {
+            log.tracef("Skip locking or keys already locked. Skipping command %s", command);
          }
+         return false;
+      } else if (isLocalOnly(command)) {
+         lockOrRegisterBackupLock(ctx, command.getKeys(), getLockTimeoutMillis(command));
+         ctx.addAffectedKey(command.getKeys());
+         return invokeNext(ctx, command);
+      } else {
+         return ctx.isOriginLocal() ?
+                handleLocalLockCommand(ctx, command) :
+                handleRemoteLockCommand(ctx, command);
       }
+   }
 
-      return invokeNextThenApply(ctx, command, localLockCommandWork);
+   private boolean readNeedsLock(InvocationContext ctx, FlagAffectedCommand command) {
+      return ctx.isInTxScope() &&
+             ctx.isOriginLocal() &&
+             command.hasAnyFlag(FlagBitSets.FORCE_WRITE_LOCK) &&
```

Comment:
`command.getFlagBitSet() & (FlagBitSets.FORCE_WRITE_LOCK | FlagBitSets.SKIP_LOCKING) == FlagBitSets.FORCE_WRITE_LOCK`
Maybe we should throw exception when both flags are used together.